oonib/testhelpers/http_helpers.py

```diff
@@ -1,16 +1,87 @@
+import os
 import json
 import random
 import string
+import tempfile
+from hashlib import sha256
 
-from twisted.internet import protocol, defer
+from urlparse import urlparse
 
-from cyclone.web import RequestHandler, Application
+from twisted.python.failure import Failure
+
+from twisted.internet.protocol import Factory, Protocol
+from twisted.internet.endpoints import TCP4ClientEndpoint
+
+from twisted.internet.error import DNSLookupError, TimeoutError
+from twisted.internet.error import ConnectionRefusedError
+from twisted.internet import protocol, defer, reactor
+
+from twisted.names.error import DNSNameError, DNSServerError
+from twisted.names import client as dns_client
+from twisted.names import dns
+
+from twisted.web._newclient import HTTPClientParser, ParseError
+from twisted.web.client import Agent, BrowserLikeRedirectAgent, readBody
+from twisted.web.client import ContentDecoderAgent, GzipDecoder
+from twisted.web.http_headers import Headers
+from twisted.web import error
+
+from twisted.web._newclient import ResponseFailed
+
+from cyclone.web import RequestHandler, Application, HTTPError
+from cyclone.web import asynchronous
 
 from twisted.protocols import policies, basic
 from twisted.web.http import Request
 
 from oonib import log, randomStr
 
+# Monkey patch to overcome issue in parsing of HTTP responses that don't
+# contain 3 parts in the HTTP response line
+old_status_received = HTTPClientParser.statusReceived
+def statusReceivedPatched(self, status):
+    try:
+        return old_status_received(self, status)
+    except ParseError as exc:
+        if exc.args[0] == 'wrong number of parts':
+            return old_status_received(self, status + " XXX")
+        raise
+HTTPClientParser.statusReceived = statusReceivedPatched
+
+class FixedRedirectAgent(BrowserLikeRedirectAgent):
```

Comment:
Interesting to see that the addition of web connectivity and the addition of logic on the server side lead to (increased?) code sharing between ooni-probe and ooni-backend!